File: mns-reactor/src/main/java/cz/fb/manaus/reactor/price/FairnessPolynomialCalculator.java
package cz.fb.manaus.reactor.price;

import cz.fb.manaus.core.model.MarketPrices;
import cz.fb.manaus.core.model.Side;
import org.apache.commons.math3.analysis.polynomials.PolynomialFunction;
import org.apache.commons.math3.analysis.solvers.LaguerreSolver;
import org.apache.commons.math3.exception.NoBracketingException;
import org.springframework.stereotype.Component;

import java.util.LinkedList;
import java.util.List;
import java.util.OptionalDouble;

import static java.util.stream.Collectors.toList;

@Component
public class FairnessPolynomialCalculator {

    private PolynomialFunction toPolynomial(double price) {
        return new PolynomialFunction(new double[]{1, price - 1});
    }

    public OptionalDouble getFairness(double winnerCount, List<OptionalDouble> prices) {
        if (prices.stream().allMatch(OptionalDouble::isPresent)) {
            var presentPrices = prices.stream().map(OptionalDouble::getAsDouble).collect(toList());
            var rightSide = multiplyPolynomials(presentPrices);
            rightSide = rightSide.multiply(new PolynomialFunction(new double[]{winnerCount}));

            var leftSideItems = new LinkedList<PolynomialFunction>();
            for (var i = 0; i < presentPrices.size(); i++) {
                var otherPrices = new LinkedList<Double>();
                for (var j = 0; j < presentPrices.size(); j++) {
                    if (i != j) {
                        otherPrices.add(presentPrices.get(j));
                    }
                }
                leftSideItems.add(multiplyPolynomials(otherPrices));
            }
            var leftSide = leftSideItems.stream().reduce(PolynomialFunction::add).get();
            var equation = leftSide.subtract(rightSide);

            var laguerreSolver = new LaguerreSolver();
            try {
                var root = laguerreSolver.solve(100, equation, 0, 1000);
                return OptionalDouble.of(1 / root);
            } catch (NoBracketingException exception) {
                return OptionalDouble.empty();
            }
        } else {
            return OptionalDouble.empty();
        }
    }

    public Fairness getFairness(MarketPrices marketPrices) {
        return new Fairness(
                getFairness(marketPrices.getWinnerCount(), marketPrices.getBestPrices(Side.BACK)),
                getFairness(marketPrices.getWinnerCount(), marketPrices.getBestPrices(Side.LAY)));
    }

    private PolynomialFunction multiplyPolynomials(List<Double> prices) {
        List<PolynomialFunction> polynomials = prices.stream().map(this::toPolynomial).collect(toList());
        return polynomials.stream().reduce(PolynomialFunction::multiply).get();
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/price/FairnessPolynomialCalculator.kt
package cz.fb.manaus.reactor.price

import cz.fb.manaus.core.model.RunnerPrices
import cz.fb.manaus.core.model.Side
import org.apache.commons.math3.analysis.polynomials.PolynomialFunction
import org.apache.commons.math3.analysis.solvers.LaguerreSolver
import org.apache.commons.math3.exception.NoBracketingException

object FairnessPolynomialCalculator {

    private fun toPolynomial(price: Double): PolynomialFunction {
        return PolynomialFunction(doubleArrayOf(1.0, price - 1))
    }

    fun getFairness(winnerCount: Int, prices: List<Double?>): Double? {
        if (prices.all { it != null }) {
            val presentPrices = prices.mapNotNull { it }
            val rightSide = multiplyPolynomials(presentPrices)
                .multiply(PolynomialFunction(doubleArrayOf(winnerCount.toDouble())))

            val leftSideItems = mutableListOf<PolynomialFunction>()
            for (i in presentPrices.indices) {
                val otherPrices = mutableListOf<Double>()
                for (j in presentPrices.indices) {
                    if (i != j) {
                        otherPrices.add(presentPrices[j])
                    }
                }
                leftSideItems.add(multiplyPolynomials(otherPrices))
            }
            val leftSide = leftSideItems.reduce { obj, p -> obj.add(p) }
            val equation = leftSide.subtract(rightSide)

            val laguerreSolver = LaguerreSolver()
            return try {
                val root = laguerreSolver.solve(100, equation, 0.0, 1000.0)
                1 / root
            } catch (exception: NoBracketingException) {
                null
            }
        } else {
            return null
        }
    }

    fun getFairness(marketPrices: List<RunnerPrices>): Fairness {
        val back = getFairness(1, getBestPrices(marketPrices, Side.BACK))
        val lay = getFairness(1, getBestPrices(marketPrices, Side.LAY))
        return Fairness(back = back, lay = lay)
    }

    private fun multiplyPolynomials(prices: List<Double>): PolynomialFunction {
        val polynomials = prices.map { this.toPolynomial(it) }
        return polynomials.reduce { obj, p -> obj.multiply(p) }
    }

}
